### PR TITLE
[KCCM - service controller]: KEP-3458 implementation

### PR DESCRIPTION
**Pull request by alexanderConstantinescu**
**TL;DR**

This patch implements: https://github.com/kubernetes/enhancements/issues/3458

It implements exactly what was agreed on the KEP, that is to say: 

* we only enqueue node changes to the "exclusion annotation" 
* We always list and use all nodes which are not excluded and don't have the `ToBeDeleted...` taint

**Background:**

The cluster-autoscaler (CA) taints nodes which are to be downscaled and deleted by it. The CA downscales nodes and removes them, [following this sequence of events](https://github.com/kubernetes/autoscaler/issues/5201#issuecomment-1282741177):

1. Taint the node
2. Drain the node of pods
3. Once drain is completed: delete VM
4. The node_lifecycle_controller will then delete the `Node` object, when it realizes that the VM doesn't exist anymore.

We decided to keep `ToBeDeleted...` as a predicate. We did this simply because we want to avoid issues whereby the node_lifecycle_controller is slow to react to the VM deletion and the service controller happens to process a service/node related event in the meantime. This would have caused reconciliation errors in the service controller, because we would have continued passing the Node object for which the VM now has been deleted, presumably causing the cloud provider code to error when it either a) attempts to retrieve the VM b) adds the VM reference to the load balancer set. 

/cc @thockin       

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

* Enabling connection draining for terminating pods on downscaled nodes 
* Cleaning up the KCCM logic 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
[KCCM - service controller]: enable connection draining for terminating pods upon node downscale by the cluster autoscaler. This is done by not reacting to the taint used by the cluster autoscaler to indicate that the node is going away soon, thus keeping the node referenced by the load balancer until the VM has been completely deleted.    
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig network
/sig cloud-provider

### CODE REVIEW

**Comment by thockin** [comment]
So we believe this is good and safe on  GCP, Azure, and AWS.

The scary thing about how cluster autoscaler works today is that is has quietly assumed this would be a non-issue all along.  If there's any cloud in which IS NOT safe, cluster autoscaler works by winning a race condition.

This PR just makes that race window 0.  CA will never win :)

So it would be super if a) someone from AWS side and Azure side can confirm; and b) people from other providers could weigh in.

**Comment by thockin** [comment]
@alexanderConstantinescu what do you think about putting this behind a feature gate?  We can start in beta (there's no API to rollback), but that way we know there's an "off switch" if we did impact someone - we'll never REALLY know until we roll it out.

Then if we pass a release or 2 we can move it to GA.

Usually a gate means a KEP - am I asking too much of you for this one little thing?  It can be a small KEP.  I can help...

**Comment by aojea** [comment]
You can follow the Sig-node exec timeout KEP, it is similar to this problem, is just a small KEP to allow to rollout changes safely

**Comment by thockin** [comment]
This one?
https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1972-kubelet-exec-probe-timeouts


"GA, on by default, but disableable" is the same as what we call "beta"
IMO, so sure :)

**Comment by aojea** [comment]
exactly

**Comment by alexanderConstantinescu** [comment]
I don't have any objections...however, https://github.com/kubernetes/enhancements/pull/3460 already mentions this problem already. Do you prefer that we deal with it on that KEP or that a file a dedicated one for this problem? I will work on refining the KEP this week as to make it a bit more explicit w.r.t what it's trying to solve. Given that we're in the middle of the 1.27 KEP cycle: merging this PR could also be postponed to after that KEP?

**Comment by thockin** [comment]
#3460 has a LOT of stuff in it (proposed, anyway).  I think it's actually 3
or 4 different KEPs.

1) LB set stability (highest prio) - could include this topic (ignoring the
CA taint )
2) kube-proxy better reporting of healthiness (high prio)
3) kube-proxy HC serving reliability (lower prio)
4) kube-proxy apiserver "majority report" (lower prio)

File: staging/src/k8s.io/controller-manager/pkg/features/kube_features.go [hunk 1]
```diff
@@ -47,5 +55,6 @@ func SetupCurrentKubernetesSpecificFeatureGates(featuregates featuregate.Mutable
 // cloudPublicFeatureGates consists of cloud-specific feature keys.
 // To add a new feature, define a key for it at k8s.io/api/pkg/features and add it here.
 var cloudPublicFeatureGates = map[featuregate.Feature]featuregate.FeatureSpec{
-	IPv6DualStack: {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
```
Comment (aojea) [inline staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:50]: this was removed from the core, @cheftako @andrewsykim can we remove it here too?

Comment (aojea) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:50]: I see it is removed in next commit, but let's give visibility to the sig-cloud-provider owners

File: staging/src/k8s.io/cloud-provider/options/options.go [hunk 2]
```diff
@@ -42,9 +42,6 @@ import (
 	cmoptions "k8s.io/controller-manager/options"
 	"k8s.io/controller-manager/pkg/clientbuilder"
 	netutils "k8s.io/utils/net"
-
-	// add the related feature gates
```
Comment (aojea) [inline staging/src/k8s.io/cloud-provider/options/options.go:46]: maybe they want to keep it for later? better ask Andrew and Walter before removing

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go [hunk 3]
```diff
@@ -658,6 +660,9 @@ func nodeNames(nodes []*v1.Node) sets.String {
 }
 
 func shouldSyncUpdatedNode(oldNode, newNode *v1.Node) bool {
+	if utilfeature.DefaultFeatureGate.Enabled(features.StableLoadBalancerNodeSet) {
```
Comment (aojea) [inline staging/src/k8s.io/cloud-provider/controllers/service/controller.go:663]: this should have a comment explaining it 😄 if you are not familiar with this is impossible you can infer what this does :

**Comment by aojea** [comment]
/assign @cheftako @andrewsykim 

cloud-providers have their own feature gates, I really don't know if is ok to remove them in this PR, maybe they need to use a different set than the kubernetes one?

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go (hunk 3 above)
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:663]: I can add a short comment mentioning: "this is intended to reduce the number of re-configurations for the load balancer node set". But honestly, there's an entire KEP for people to read if they are interested in understanding the "why"?

File: staging/src/k8s.io/cloud-provider/controllers/service/metrics.go [hunk 4]
```diff
@@ -32,12 +33,26 @@ var register sync.Once
...
+		legacyregistry.MustRegister(loadBalancerSyncCount)
 		legacyregistry.MustRegister(nodeSyncLatency)
+		legacyregistry.MustRegister(nodeSyncErrorCount)
 		legacyregistry.MustRegister(updateLoadBalancerHostLatency)
 	})
 }
 
 var (
+	loadBalancerSyncCount = metrics.NewCounter(&metrics.CounterOpts{
+		Name:           "load_balancer_sync_count",
```
Comment (logicalhan) [inline staging/src/k8s.io/cloud-provider/controllers/service/metrics.go:45]: counters should be prefixed as `_total`.

File: staging/src/k8s.io/cloud-provider/controllers/service/metrics.go [hunk 5]
```diff
@@ -32,12 +33,26 @@ var register sync.Once
...
+		Subsystem:      subSystemName,
+		Help:           "A metric counting the amount of times any load balancer has been configured, as an effect of service/node changes on the cluster",
+		StabilityLevel: metrics.BETA,
+	})
+	nodeSyncErrorCount = metrics.NewCounter(&metrics.CounterOpts{
+		Name:           "node_sync_error_count",
```
Comment (logicalhan) [inline staging/src/k8s.io/cloud-provider/controllers/service/metrics.go:51]: same.

**Review by logicalhan — COMMENTED** [review]
/triage accepted

File: staging/src/k8s.io/cloud-provider/controllers/service/metrics.go (hunk 4 above)
Comment (thockin) [reply staging/src/k8s.io/cloud-provider/controllers/service/metrics.go:45]: suffxed?

Comment (logicalhan) [reply staging/src/k8s.io/cloud-provider/controllers/service/metrics.go:45]: Hahaha yeah 😂

**Review by thockin — COMMENTED** [review]
I didn't dig into all the test cases yet, this is a nice small PR except for that :)

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go (hunk 3 above)
Comment (aojea) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:663]: I don't mean to be exhaust, but what is the expected output of this ? something as "only nodes with ... will be updated"

```
respectsPredicates(oldNode, nodeIncludedPredicate) != respectsPredicates(newNode, nodeIncludedPredicate)
```

File: pkg/features/kube_features.go [hunk 6]
```diff
@@ -1029,6 +1037,8 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	SizeMemoryBackedVolumes: {Default: true, PreRelease: featuregate.Beta},
 
+	StableLoadBalancerNodeSet: {Default: true, PreRelease: featuregate.Beta},
```
Comment (aojea) [inline pkg/features/kube_features.go:1040]: gce is the main stakeholder, I would say the only one of this change, I don't think we should rollout this change enable by default @thockin ... there is not testing at all in the fleet for this

Comment (aojea) [reply pkg/features/kube_features.go:1040]: I thought I commented in the KEP about this but I can't find the comment, I see that it merged with beta as default 🤷 , we are going YOLO then

Comment (alexanderConstantinescu) [reply pkg/features/kube_features.go:1040]: > gce is the main stakeholder (the only one) of this change 

No, you're confusing this with "KEP 2", the one changing Kube-proxy's `healthz` to also inspect the Node state - probably the `unschedulable` field - when determining if it should be healthy or not. That impacts GCP because GCP is the only one targeting Kube-proxy's healthz for eTP:Cluster services. 

This applies to all cloud providers equally.

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go [hunk 7]
```diff
@@ -932,9 +937,16 @@ var (
 		nodeIncludedPredicate,
 		nodeUnTaintedPredicate,
 	}
+	stableNodeSetPredicates []NodeConditionPredicate = []NodeConditionPredicate{
+		nodeIncludedPredicate,
+		nodeUnTaintedPredicate,
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:944]: I thought we landed on not syncing load balancers as a consequence of the addition/deletion of this taint, but making sure we never add nodes to the LB set which might have this taint, see: https://github.com/kubernetes/enhancements/pull/3460#discussion_r1095135706 

Logically, what you mentioned is true and possible, though maybe not likely. 

> If CA taints a Node, then deletes the VM, and NodeController is slow (say it takes 60 seconds to wake up), any subsequent LB set changes will include this Node, but the VM for it is already gone, so will fail.

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go [hunk 8]
```diff
@@ -932,9 +937,16 @@ var (
 		nodeIncludedPredicate,
 		nodeUnTaintedPredicate,
 	}
+	stableNodeSetPredicates []NodeConditionPredicate = []NodeConditionPredicate{
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:942]: Yeah, I did this because I wanted the code to read better should something go wrong with this feature and some developer starts going through the code to debug. Keeping `etpLocalNodePredicates` and `stableNodeSetPredicates` separate, was my intention of indicating what has changed w.r.t this feature and what was there before (i.e: that these were already the applied predicates for eTP:Local).    

This might not be a valid use case?

File: staging/src/k8s.io/controller-manager/pkg/features/kube_features.go [hunk 9]
```diff
@@ -48,13 +34,12 @@ const (
 	StableLoadBalancerNodeSet featuregate.Feature = "StableLoadBalancerNodeSet"
 )
 
-func SetupCurrentKubernetesSpecificFeatureGates(featuregates featuregate.MutableFeatureGate) error {
-	return featuregates.Add(cloudPublicFeatureGates)
+func init() {
+	runtime.Must(utilfeature.DefaultMutableFeatureGate.Add(cloudPublicFeatureGates))
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:38]: > Why is it in this PR?

As it was: `SetupCurrentKubernetesSpecificFeatureGates` was only called from the init function in  [`package register`](https://github.com/kubernetes/kubernetes/blob/b3d8ac8496a23d65a907f9333d906bcd5463764e/staging/src/k8s.io/controller-manager/pkg/features/register/register.go), forcing any test file to reference that package as to test code depending on the feature gate. Right now, this follows the general convention across the code base and forces test files to just call [`package features`](https://github.com/kubernetes/kubernetes/blob/1984eeff3bc39fd0595099a6ce2d5b686ee064a9/staging/src/k8s.io/controller-manager/pkg/features/kube_features.go) - which needs to happen in any case when accessing the variable `StableLoadBalancerNodeSet`. 

This was never picked up on before, because the feature gate `IPv6DualStack` was never tested from the controller-manager.

File: staging/src/k8s.io/controller-manager/pkg/features/kube_features.go [hunk 10]
```diff
@@ -17,35 +17,29 @@ limitations under the License.
...
 	"k8s.io/component-base/featuregate"
 )
 
 const (
 	// Every feature gate should add method here following this template:
+	// owner: @alexanderConstantinescu
+	// kep: http://kep.k8s.io/3458
+	// beta: v1.27
 	//
-	// // owner: @username
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:26]: By mistake...

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go (hunk 3 above)
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:663]: >  but what is the expected output of this ? 

Only Nodes with changes to the label `"node.kubernetes.io/exclude-from-external-load-balancers"` will trigger a load balancer re-sync.

Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:663]: Want me to add that as a doc bloc?

Comment (aojea) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:663]: yeah please, if someone like me has to debug something in this area, that will give me a hint, otherwise I have to do `git blame, read KEP, ...`

```
// only Nodes with changes to the label "node.kubernetes.io/exclude-from-external-load-balancers" will trigger a load balancer re-sync.
```

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (thockin) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:945]: I'm OK either way.  I see your point.

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (thockin) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:947]: > I thought we landed on not syncing load balancers as a consequence of the addition/deletion of this taint, but making sure we never add nodes to the LB set which might have this taint

Agree, but is this what is happening?  The informer setup code filters node updates, but if a node is added or removed, we sync all LBs, and in doing so we get all nodes and then filter by stableNodeSetPredicates, which includes the taint.

So chanign the taint won't trigger a sync, but any node being added or removed would cause all taineted nodes to be removed?  I don't think that's what we want, is it?

[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:947]: >  I don't think that's what we want, is it?

Are you saying this because we wouldn't connection drain the tainted nodes? I think we can live with that, simply because I don't think this corner case has a very large window of opportunity to occur....and if it does occur: the effect is that we don't connection drain, which we can live with.   

Beyond what you mentioned: 

> If CA taints a Node, then deletes the VM, and NodeController is slow (say it takes 60 seconds to wake up), any subsequent LB set changes will include this Node, but the VM for it is already gone, so will fail.

the events triggering "subsequent LB set changes", which would be a) the addition of a new Node b) change to `"node.kubernetes.io/exclude-from-external-load-balancers"` c) the deletion of a Node but not by the NodeController (so, only by a user because the node controller is the only thing in Kube deleting nodes, I believe). 
In "KEP-2" kube-proxy will start connection draining on the `unschedulable` condition: reducing the window of opportunity to not connection drain even further. 

Also

>  I mentioned excluding a tainted node if it was not already in the LB set,

Any tainted node will always be in the LB set the moment it is tainted. So if we're only excluding the tainted node if it's not in the LB set, then we're probably never excluding it, right? 

What I am trying to say is: I think not syncing on changes to the taint but excluding all nodes with it, is fine. I am also fine with removing the taint altogether, because I think it's acceptable to have the service controller experience transient errors until this condition resolves. But I can also see the argument of people operating the control plane components and wanting to avoid errors

File: staging/src/k8s.io/controller-manager/pkg/features/kube_features.go [hunk 11]
```diff
@@ -17,35 +17,39 @@ limitations under the License.
...
 	// Feature gates should be listed in alphabetical, case-sensitive
 	// (upper before any lower case character) order. This reduces the risk
 	// of code conflicts because changes are more likely to be scattered
 	// across the file.
 
-	// owner: @khenidak
-	// alpha: v1.15
+	// owner: @alexanderConstantinescu
+	// kep: http://kep.k8s.io/3458
+	// beta: v1.27
```
Comment (andrewsykim) [inline staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:40]: Sorry if I missed this in the KEP, but what's the reason for introducing this feature as Beta and enabled by default?

File: staging/src/k8s.io/controller-manager/pkg/features/kube_features.go (hunk 1 above)
Comment (andrewsykim) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:50]: Can we do this feature gate clean up in a separate PR?

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (thockin) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:947]: >> I don't think that's what we want, is it?
>
> Are you saying this because we wouldn't connection drain the tainted nodes?

Exactly.  In a cluster that is scaling down via CA, will we only see one of these at a time?  If so, it's probably OK.

My fear is that something like:

* LB set is nodes A, B, C
* CA decides to "rebalance" or make 2 independent decisions.  E.g. "node A is almost empty, scale down" and "there's a pod that needs some special resource which needs a new node"
* taint Node A (but do not remove it) & set unschedulable
* draining period starts
* new node comes up
* node A is removed from LB set

How about this:  Add a big fat comment here, along the lines of:

```
// This is not perfect, but probably good enough.  We won't update the LBs
// just because the taint was added (see shouldSyncUpdatedNode) but if any
// other situation causes an LB sync, tainted nodes will be excluded at that
// time.
```

To do better than that is actually more complicated - just leaving tainted nodes that are already in a set can still fail (as you cited me saying :), and the corner-case is small.  Just leave us a clue so when customers call about occasional errors from LBs, we can hate past-us a little less. :)

File: staging/src/k8s.io/controller-manager/pkg/features/kube_features.go (hunk 11 above)
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:40]: The genesis to this lies here: https://github.com/kubernetes/kubernetes/pull/115204#issuecomment-1399095942

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:947]: OK, I'll update with that comment.

File: staging/src/k8s.io/controller-manager/pkg/features/kube_features.go (hunk 11 above)
Comment (andrewsykim) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:40]: Ah I see the comments earlier in the PR now discussing this.

File: staging/src/k8s.io/controller-manager/pkg/features/kube_features.go (hunk 1 above)
Comment (thockin) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:50]: +1

File: staging/src/k8s.io/cloud-provider/options/options.go (hunk 2 above)
Comment (thockin) [reply staging/src/k8s.io/cloud-provider/options/options.go:46]: Did we get a decision on this?

File: staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go [hunk 12]
```diff
@@ -2334,26 +2508,33 @@ func Test_shouldSyncUpdatedNode_individualPredicates(t *testing.T) {
 					},
 				},
 			},
-			shouldSync: false,
+			shouldSync:           true,
+			stableNodeSetEnabled: true,
 		},
 		{
-			name: "readiness changed F->T",
+			name: "excluded changed T->T",
```
[the remark this answers is not in the document]
Comment (thockin) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:2515]: I did the following cleanup (which I can send as a PR after this one merges, or you can steal and apply it yourself): https://github.com/kubernetes/kubernetes/commit/3d63aed2dca23bfec17c174b32b63c0be511fe85

This makes it clear what is happening.  It also shows obviously redundant tests, fixed in https://github.com/kubernetes/kubernetes/commit/b5e4bbb132d0d6c4079e5c98a3bb8ab6e88be381

Resolving this comment

File: staging/src/k8s.io/controller-manager/pkg/features/kube_features.go (hunk 1 above)
Comment (thockin) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:50]: Other than breaking these unrelated bits out, I am LGTM on this PR

Comment (alexanderConstantinescu) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:50]: OK, I will remove the second commit in that case.

File: staging/src/k8s.io/cloud-provider/options/options.go (hunk 2 above)
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/options/options.go:46]: This is what https://github.com/kubernetes/kubernetes/pull/115204/commits/8360f9c0302f6edfdd5b13bbe69f7f283a1d4742#r1107217918 talks about, right? So I understand we the decision is to remove the commit from this PR

File: staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:2515]: > which I can send as a PR after this one merges

Sounds good, please do that then. Sorry for the confusion with these tests.

File: staging/src/k8s.io/controller-manager/pkg/features/kube_features.go (hunk 1 above)
Comment (andrewsykim) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:50]: FYI: https://github.com/kubernetes/kubernetes/pull/116255

Comment (danwinship) [reply staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:50]: yeah, I came across this while adding a feature gate for a different KEP. 116255 is in the merge queue now

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve
/retest

**Comment by alexanderConstantinescu** [comment]
FYI: the bot might have a bug, it missed adding approve label even though [`/approve` ](https://github.com/kubernetes/kubernetes/pull/115204#issuecomment-1457190980)was spelled correctly  

Can someone retry?

**Comment by aojea** [comment]
you need approval from https://github.com/kubernetes/kubernetes/blob/master/test/instrumentation/testdata/OWNERS

**Comment by dims** [comment]
/assign @serathius @logicalhan @ehashman

(for `test/instrumentation/testdata/` approval)

**Comment by thockin** [comment]
It's somewhat reassuring that there are things I can't approve :)

File: staging/src/k8s.io/cloud-provider/controllers/service/metrics.go [hunk 13]
```diff
@@ -32,12 +33,26 @@ var register sync.Once
...
 		legacyregistry.MustRegister(updateLoadBalancerHostLatency)
 	})
 }
 
 var (
+	loadBalancerSyncCount = metrics.NewCounter(&metrics.CounterOpts{
+		Name:           "load_balancer_sync_total",
+		Subsystem:      subSystemName,
+		Help:           "A metric counting the amount of times any load balancer has been configured, as an effect of service/node changes on the cluster",
+		StabilityLevel: metrics.BETA,
```
Comment (serathius) [inline staging/src/k8s.io/cloud-provider/controllers/service/metrics.go:48]: Feature stability != Metric Stability.
Feature being in beta doesn't meant it's metrics need to be in Beta stage. 
Beta metrics have it's own special meaning https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3498-extending-stability#beta-metrics

I would recommend to start with Alpha metric stage. Those metrics don't need to registered in test/instrumentation/testdata/stable-metrics-list.yaml and be reviewed by SIG-instrumentation.

**Comment by thockin** [comment]
@alexanderConstantinescu ping me when ready

**Comment by alexanderConstantinescu** [comment]
> @alexanderConstantinescu ping me when ready

@thockin : updated the PR and changed the metrics to `stabilityLevel: ALPHA`

File: pkg/features/kube_features.go [hunk 14]
```diff
@@ -739,6 +739,14 @@ const (
 	// Enables kubelet support to size memory backed volumes
 	SizeMemoryBackedVolumes featuregate.Feature = "SizeMemoryBackedVolumes"
 
+	// owner: @alexanderConstantinescu
+	// kep: http://kep.k8s.io/3458
+	// beta: v1.27
+	//
+	// Enables less load balancer re-configurations by the service controller
+	// (KCCM) as an effect of changing node state.
+	StableLoadBalancerNodeSet featuregate.Feature = "StableLoadBalancerNodeSet"
```
Comment (aojea) [inline pkg/features/kube_features.go:748]: do we need this here?
It doesn't seem we need it here since the cloud-provider has its own features gates and are the ones that we are importing

File: staging/src/k8s.io/cloud-provider/controllers/service/metrics.go [hunk 15]
```diff
@@ -32,12 +33,26 @@ var register sync.Once
...
 
 var (
+	loadBalancerSyncCount = metrics.NewCounter(&metrics.CounterOpts{
+		Name:           "load_balancer_sync_total",
+		Subsystem:      subSystemName,
+		Help:           "A metric counting the amount of times any load balancer has been configured, as an effect of service/node changes on the cluster",
+		StabilityLevel: metrics.ALPHA,
+	})
+	nodeSyncErrorCount = metrics.NewCounter(&metrics.CounterOpts{
+		Name:           "node_sync_error_total",
```
Comment (serathius) [inline staging/src/k8s.io/cloud-provider/controllers/service/metrics.go:51]: Below metric writes "nodesync" together. I would recommend to keep it consistent either use `nodesync_error_total` or also update `nodesync_latency_seconds` to `node_sync_latency_seconds`. Those are Alpha metrics so rename is not a problem as long as it mentioned in changelog.

File: pkg/features/kube_features.go (hunk 14 above)
Comment (alexanderConstantinescu) [reply pkg/features/kube_features.go:748]: I don't know. I thought this was the centralized location for all feature gates in kube and staging/src/k8s.io/controller-manager/pkg/features/kube_features.go the feature gates for the controller-manager, which can't import and reference `pkg/featuress/kube_features.go`

File: staging/src/k8s.io/cloud-provider/controllers/service/metrics.go (hunk 15 above)
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/controllers/service/metrics.go]: OK, I went with `nodesync_error_total ` as to not break any user/consumer depending on or using `nodesync_latency_seconds`

File: pkg/features/kube_features.go (hunk 14 above)
Comment (thockin) [reply pkg/features/kube_features.go:748]: I admit I don't know the protocol here.  pkg/features/kube_features.go obviously governs things in the k/k Go module, but since staging repos are "independent" (just pretend :), they can't rely on k/k/pkg/....

File: staging/src/k8s.io/cloud-provider/controllers/service/metrics.go [hunk 16]
```diff
@@ -32,12 +33,26 @@ var register sync.Once
...
+		legacyregistry.MustRegister(loadBalancerSyncCount)
 		legacyregistry.MustRegister(nodeSyncLatency)
+		legacyregistry.MustRegister(nodeSyncErrorCount)
 		legacyregistry.MustRegister(updateLoadBalancerHostLatency)
 	})
 }
 
 var (
+	loadBalancerSyncCount = metrics.NewCounter(&metrics.CounterOpts{
+		Name:           "load_balancer_sync_total",
```
Comment (thockin) [inline staging/src/k8s.io/cloud-provider/controllers/service/metrics.go:45]: super nit:  below we have "loadbalancer" and here we have "load_balancer".  Since the "update_loadbalancer_host_latency_seconds" let's go with "loadbalancer" ?

I admit to being kind of annoyed at "nodesync" vs "loadbalancer_sync", but oh well :)

**Comment by thockin** [comment]
I am happy - @serathius?

/lgtm
/hold

**Comment by serathius** [comment]
> I am happy - @serathius?

Too

/lgtm
/unhold
